clang/include/clang/AST/ASTContext.h

```diff
@@ -2803,6 +2803,10 @@ class ASTContext : public RefCountedBase<ASTContext> {
     return getUnqualifiedArrayType(T, Quals);
   }
 
+  // Determine whether an array is a valid return type
+  // Array is a valid return type for HLSL
+  bool isReturnableArrayType() const { return getLangOpts().HLSL; }
```

Comment:
This should probably also take the type and check that the type is an array type, otherwise it's more of a language options check for whether or not array types are allowed to be returned.